### PR TITLE
Mapping merge driver: continued

### PR DESCRIPTION
**Pull request by 20kdc**
A continuation of DrSmugleaf's PR #1478

### CODE REVIEW

**Comment by 20kdc** [comment]
To clarify, status is "not fully tested in all cases", so I'm going to need to know what to add next.
It also may potentially eat changes in some cases involving adjustments to sequenced YAML. Sane schemaless structural merging remains nigh-impossible.

**Comment by 20kdc** [comment]
Ok, so the big problem with this is: Map files keep having their UIDs changed around
The mapping merge driver has no way whatsoever of fixing this

File: Content.Tools/Merger.cs [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+                // Console.WriteLine("Entity C/" + kvp.Key + " -> A/" + kvp.Value);
+                YamlMappingNode oursEnt;
+                if (MapOurs.Entities.ContainsKey(kvp.Value))
+                {
+                    YamlMappingNode basedEnt;
+                    oursEnt = MapOurs.Entities[kvp.Value];
+                    if (MapBased.Entities.ContainsKey(kvp.Value))
+                    {
+                        basedEnt = MapBased.Entities[kvp.Value];
+                    }
```
Comment (metalgearsloth) [inline Content.Tools/Merger.cs:215]: Could use TryGetValue for these.

File: Content.Tools/Merger.cs [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+                    aMap[ind] = res;
+                }
+                aMap[ind].Children["tiles"] = Convert.ToBase64String(outB);
+            }
+        }
+
+        public uint MapTileId(uint src, Dictionary<uint, uint> mapping)
+        {
+            return (src & 0xFFFF0000) | mapping[src & 0xFFFF];
+        }
```
Comment (metalgearsloth) [inline Content.Tools/Merger.cs:146]: Yeah I don't know how tiles work.

File: Content.Tools/Merger.cs [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+            {
+                var name = component["type"].ToString();
+                basedComponents[name] = (YamlMappingNode) component;
+            }
+            foreach (var otherComponent in otherComponentsNode)
+            {
+                var name = otherComponent["type"].ToString();
+                if (ourComponents.ContainsKey(name))
+                {
+                    var ourComponent = ourComponents[name];
```
Comment (metalgearsloth) [inline Content.Tools/Merger.cs:271]: TryGetValue here as well.

File: Content.Tools/Merger.cs [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+                else
+                {
+                    Console.WriteLine($"Error finding UID in MapEntityRecursiveAndBadly {path}. To fix this, the merge driver needs to be improved.");
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        public bool MapEntityRecursiveAndBadly(YamlNode node, string path)
```
Comment (metalgearsloth) [inline Content.Tools/Merger.cs:337]: lol

**Review by metalgearsloth — COMMENTED** [review]
I tried. Didn't help I didn't understand why there's 3 maps involved in a merge.

We should probably merge it and not use it until we do hashing or whatever to fix the uids changing every other commit.

**Comment by 20kdc** [comment]
Ok, so before I handle each individual review, I'm going to write an explaination of how a merge driver is supposed to work.

There are two "finished" branches to merge, and the "common base".
The goal of the merge driver is to apply the changes made by both branches from the common base, inferring what those changes are, exactly, from the data in the common base.

For example, given these two hypothetical strings that you are supposed to merge:
A:`ABCDEF`
and
C:`ABCXDG`
you can't' reliably merge these without ignoring any deletions that were intentionally made in one branch and not another.
But if you know of the common base string, B:`ABCDE`, then you know that A adds 'F', while C removes 'D' and adds 'G'.
Therefore, you know the correct answer is `ABCEFG` or `ABCEGF`.
This is the basic principle behind why a merge driver must receive 3 sources - it uses the common base to determine what changes each branch made.

File: Content.Tools/Merger.cs [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+        public List<uint> EntityListDirectMerge { get; } = new List<uint>();
+
+        private const int ExpectedChunkSize = 16 * 16 * 4;
+
+        public Merger(Map ours, Map based, Map other)
+        {
+            MapOurs = ours;
+            MapBased = based;
+            MapOther = other;
+        }
```
[the remark this answers is not in the document]
Comment (20kdc) [reply Content.Tools/Merger.cs:29]: Please see the description of the operation of a merge driver I put as a comment on the PR.

File: Content.Tools/Merger.cs [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+                        if (aI == bI)
+                        {
+                            // If aI == bI then aI did not change anything, so cI always wins
+                            result = cI;
+                        }
+                        else if (bI != cI)
+                        {
+                            // If bI != cI then cI definitely changed something (conflict, but overrides aI)
+                            result = cI;
+                            Console.WriteLine("WARNING: Tile (" + ind + ")[" + i + "] was changed by both branches.");
```
[the remark this answers is not in the document]
Comment (20kdc) [reply Content.Tools/Merger.cs:126]: To be logged where?

**Comment by PaulRitter** [comment]
@20kdc to fix the uid problem, you could write a component called `MapIDComponent` which just wraps a uint value. Add the comp to every entity when saving the map & you have a consistent uid, since the component will be read & written again every time. This will make mapfiles *slightly* larger but its worth it imo

**Comment by PJB3005** [comment]
> @20kdc to fix the uid problem, you could write a component called `MapIDComponent` which just wraps a uint value. Add the comp to every entity when saving the map & you have a consistent uid, since the component will be read & written again every time. This will make mapfiles _slightly_ larger but its worth it imo

I'll work on this if you don't mind.

**Comment by 20kdc** [comment]
I don't mind, feel free. I assume you have a good plan anyway. Just tell me when it's done and I'll get everything ready on this end.

**Comment by PJB3005** [comment]
https://github.com/space-wizards/RobustToolbox/commit/ad4c8be132bd6ee9d58d2dde0ac1e0aa524df27e adds preservation engine side. 9d5e85ba1ea22eb31ab90685d5b88e84e4b71abf is content side.

Basically it just tries to keep the `uid` of entities consistent between map saves, as long as the entity isn't like deleted or recreated.

**Comment by 20kdc** [comment]
If the implementation method used here preserves the actual `uid` in the map file according to the original format, then there's nothing that needs to be done to add further compatibility (the merge driver already utilizes UIDs in this way)

File: Content.Tools/Merger.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (PJB3005) [reply Content.Tools/Merger.cs:126]: Eh this probably is too minor to be worth bothering with.

**Pull request merged**